sync-core/src/test/java/com/cloudant/sync/query/QueryFilterFieldsTest.java

```diff
@@ -0,0 +1,203 @@
+            assertThat((String) mutableBody.get("name"), is("mike"));
+            assertThat((Integer) mutableBody.get("age"), is(12));
+            assertThat((String) mutableBody.get("pet"), is("cat"));
+        }
+    }
+
+    @Test
+    public void returnsNullMutableCopyWhenDocUpdated() {
+        // query - { "name" : "mike", "age" : 12 }
+        Map<String, Object> query = new HashMap<String, Object>();
+        query.put("name", "mike");
+        query.put("age", 12);
+        QueryResult queryResult = im.find(query, 0, Long.MAX_VALUE, Arrays.asList("name"), null);
+        assertThat(queryResult.size(), is(1));
+        for (DocumentRevision rev : queryResult) {
+            assertThat(rev, is(instanceOf(ProjectedDocumentRevision.class)));
+            Map<String, Object> revBody = rev.getBody().asMap();
+            assertThat(revBody.keySet(), contains("name"));
+            assertThat((String) revBody.get("name"), is("mike"));
+
+            BasicDocumentRevision original = ds.getDocument(rev.getId());
+            MutableDocumentRevision update = original.mutableCopy();
+            Map<String, Object> updateBody = original.getBody().asMap();
+            updateBody.put("name", "charles");
+            update.body = DocumentBodyFactory.create(updateBody);
+            try {
+                assertThat(ds.updateDocumentFromRevision(update), is(notNullValue()));
+            } catch (ConflictException e) {
+                Assert.fail("Failed to update document revision");
+                e.printStackTrace();
+            } catch (IOException e) {
+                Assert.fail("Failed to update document revision");
+                e.printStackTrace();
+            }
+
+            assertThat(((ProjectedDocumentRevision) rev).mutableCopy(), is(nullValue()));
+        }
+    }
+
+    @Test
+    public void returnsNullMutableCopyWhenDocDeleted() {
+        // query - { "name" : "mike", "age" : 12 }
+        Map<String, Object> query = new HashMap<String, Object>();
+        query.put("name", "mike");
+        query.put("age", 12);
+        QueryResult queryResult = im.find(query, 0, Long.MAX_VALUE, Arrays.asList("name"), null);
+        assertThat(queryResult.size(), is(1));
+        for (DocumentRevision rev : queryResult) {
+            assertThat(rev, is(instanceOf(ProjectedDocumentRevision.class)));
+            Map<String, Object> revBody = rev.getBody().asMap();
+            assertThat(revBody.keySet(), contains("name"));
+            assertThat((String) revBody.get("name"), is("mike"));
+
+            try {
+                assertThat(ds.deleteDocumentFromRevision((ProjectedDocumentRevision) rev),
+                        is(notNullValue()));
+            } catch (ConflictException e) {
+                Assert.fail("Failed to delete document revision");
+                e.printStackTrace();
+            }
+
+            assertThat(((ProjectedDocumentRevision) rev).mutableCopy(), is(nullValue()));
+        }
+    }
+
+}
```

Comment:
Set up IDEA to add new lines to the end of files by default (http://stackoverflow.com/questions/16761227/how-to-make-intellij-idea-insert-a-new-line-at-every-end-of-file). But it's not worth fixing this one up right now.